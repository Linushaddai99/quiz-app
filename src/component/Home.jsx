import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useEffect } from "react";
import getCategories from "../redux/categoriesApi";
import { useNavigate } from "react-router-dom";
import getQuestions from "../redux/questionsApi";

const Home = () => {
  const dispatch = useDispatch();
  const categories = useSelector((state) => state?.categories?.categories);

  useEffect(() => {
    dispatch(getCategories());
  }, [dispatch, categories?.length]);

  const navigate = useNavigate();

  const [amount, setAmount] = useState();
  const [cate, setCate] = useState();
  const [difficulty, setDifficulty] = useState();
  const [type, setType] = useState();
  const [err, setErr] = useState("");

  const formValidation = (data) => {
    const result = [];
    for (const property in data) {
      result.push(data[property]);
    }

    if (result.includes(NaN || undefined)) {
      setErr("Please Fill out all the select dropdown!!!");
    } else {
      dispatch(getQuestions(data));
      navigate("/quiz");
    }
  };

  const handleForm = (e) => {
    e.preventDefault();

    const choiceData = {
      amount: Number(amount),
      category: Number(cate),
      difficulty: difficulty,
      type: type,
    };

    formValidation(choiceData);
  };

  return (
    <div className="container selection-form">
      <h1>Take A Quiz</h1>
      <p className="error-msg">{err}</p>
      <form onSubmit={handleForm}>
        <label for="number">Number of quesions:</label>
        <input
          type="number"
          onChange={(e) => setAmount(e.target.value)}
          placeholder="Maximum of 10"
          className="text-field"
        />

        <label for="category">Select Category:</label>
        <select
          required
          name="categories"
          id="categories"
          onChange={(e) => setCate(e.target.value)}
          className="text-field"
        >
          <option value="Any Category">Choose Category</option>
          {categories?.map((category) => (
            <option value={category?.id} key={category.id}>{category?.name}</option>
          ))}
        </select>

        <label for="difficulty">Select Difficulty:</label>
        <select
          name="difficulty"
          id="difficulty"
          onChange={(e) => setDifficulty(e.target.value)}
          className="text-field"
          required
        >
          <option value="Any difficulty">Choose difficulty</option>
          <option value="easy">easy</option>
          <option value="medium">medium</option>
          <option value="hard">hard</option>
        </select>

        <label for="type">Select Type:</label>
        <select
          name="type"
          id="type"
          onChange={(e) => setType(e.target.value)}
          className="text-field"
          required
        >
          <option value="Any type">Choose type</option>
          <option value="multiple">Multiple</option>
          <option value="boolean">True/False</option>
        </select>

        <button type="submit">Generate Questions</button>
      </form>
    </div>
  );
};

export default Home;
