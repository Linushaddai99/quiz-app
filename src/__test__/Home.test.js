import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Home from '../component/Home';
import store from '../redux/configStore';

describe('Home page', () => {
  it('should display Homepage correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should render title', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByText(/TAKE A QUIZ/i)).toBeInTheDocument();
  });

  it('Should render Generate questions button', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>,
    );

    expect(screen.getByRole('button')).toBeInTheDocument();
  });
});
