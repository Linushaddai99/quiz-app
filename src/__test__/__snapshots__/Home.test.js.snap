// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should display Homepage correctly 1`] = `
<div>
  <div
    class="container selection-form"
  >
    <h1>
      Take A Quiz
    </h1>
    <form>
      <label
        for="number"
      >
        Number of quesions:
      </label>
      <input
        class="text-field"
        placeholder="Maximum of 10"
        type="number"
      />
      <label
        for="category"
      >
        Select Category:
      </label>
      <select
        class="text-field"
        id="categories"
        name="categories"
      >
        <option
          value="Any Category"
        >
          Choose Category
        </option>
      </select>
      <label
        for="difficulty"
      >
        Select Difficulty:
      </label>
      <select
        class="text-field"
        id="difficulty"
        name="difficulty"
      >
        <option
          value="Any difficulty"
        >
          Choose difficulty
        </option>
        <option
          value="easy"
        >
          easy
        </option>
        <option
          value="medium"
        >
          medium
        </option>
        <option
          value="hard"
        >
          hard
        </option>
      </select>
      <label
        for="type"
      >
        Select Type:
      </label>
      <select
        class="text-field"
        id="type"
        name="type"
      >
        <option
          value="Any type"
        >
          Choose type
        </option>
        <option
          value="multiple"
        >
          multiple
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
      </select>
      <button
        type="submit"
      >
        Generate Questions
      </button>
    </form>
  </div>
</div>
`;
